core-bundle/src/Controller/FrontendModule/TwoFactorController.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Controller\FrontendModule;
+
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Contao\CoreBundle\Security\TwoFactor\Authenticator;
+use Contao\FrontendUser;
+use Contao\ModuleModel;
+use Contao\PageModel;
+use Contao\Template;
+use ParagonIE\ConstantTime\Base32;
+use Scheb\TwoFactorBundle\Security\Authentication\Exception\InvalidTwoFactorCodeException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class TwoFactorController extends AbstractFrontendModuleController
+{
+    /** @var PageModel */
+    protected $page;
+
+    public function __invoke(Request $request, ModuleModel $model, string $section, array $classes = null, PageModel $page = null): Response
+    {
+        $this->page = $page;
+
+        if ($this->page instanceof PageModel && $this->get('contao.routing.scope_matcher')->isFrontendRequest($request)) {
+            $this->page->loadDetails();
+        }
+
+        return parent::__invoke($request, $model, $section, $classes);
+    }
+
+    public static function getSubscribedServices(): array
```

Comment:
Can you elaborate on why you did not use dependency injection? I don't think it's a good idea to use service locators. I really dislike this pattern. It might be a good choice for your local application but it looks weird to me in a public bundle. How am I supposed to replace the `translator` service for the `TwoFactorController` if I want to?

Reply:
https://symfony.com/doc/current/service_container/service_subscribers_locators.html

Service locators allow lazy loading of services, which is recommended in controllers. You can still easily override the service, they are bound to the class name that is specified, so if you use autowiring you still override the class alias to point to your correct service.

Reply:
I don't want to override an alias. I want to override just one service that's being used in this very controller.

Reply:
Using service locators in the controller seems perfectly fine to me.

Reply:
Okay, so service locators can easily be replaced via DI as well. I'm fine with it then :)